### ISSUE TITLE
Detection of correct union type based on fixed values

### ISSUE THREAD

**Issue by dmolineus** [vimeo/psalm#3548]
I tried to use a fixed value to distinguish between different types combined with the union type operator. Psalm currently doesn't detect the right type. Support such type annotations would be awesome.

```php
<?php

/** 
 * @psalm-type TCircle = array{ type: 'circle', radius: int}
 * @psalm-type TSquare = array{ type: 'square', length: int}
 * @psalm-type TGeometry = TCircle|TSquare
*/

/** @psalm-var list<TGeometry> */
$geometries = [
    ['type' => 'circle', 'radius' => 4],
    ['type' => 'square', 'length' => 4],
];

foreach ($geometries as $geometry) {
    switch ($geometry['type']) {
        case 'circle':
            echo 'Render circle with a radius of ' . $geometry['radius'] . ' pixels';
            break;
        case 'square':
            echo 'Render square with a length of ' . $geometry['length'] . ' pixels';
            break;
    }
}
```

https://psalm.dev/r/826cde9032

**Comment by weirdan** [comment]
That would be 'tagged union' or 'variant record'. Experience tells me it's not very likely to happen.

**Comment by muglug** [comment]
Yeah, not going to happen, please use classes instead.

**Comment by dmolineus** [comment]
Thanks for the feedback. The idea was for cases where it's not possible to use classes, e.g. when importing data from a external file or api.
